### PR TITLE
JPN-483 | Added function to remove certain user from Top Contributors

### PR DESCRIPTION
**Pull request by d34th4ck3r**
@Wikia/spitfires @perjg

### CODE REVIEW

File: extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php [hunk 1]
```diff
@@ -190,6 +191,47 @@ function () {
 		return $botIds;
 	}
 
+
+	/**
+	 * @return array list of blocked ids for Top Contributors
+	 */
+	private function getBlockedIds(){
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: I'd consider renaming this method. "Blocked" has a specific meaning in MediaWiki when relating to users, so when I read `getBlockedIds`, I expected the method to be getting a list of "blocked" users (i.e. blocked from editing), not a blacklist of users based on user group.

File: extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php [hunk 2]
```diff
@@ -190,6 +191,47 @@ function () {
...
+					->runLoop( $globalDb, function ( &$globalIds, $row ) {
+						$globalIds[] = $row->ug_user;
+					} );
+
+				$localDb = wfGetDB( DB_SLAVE );
+
+				$localIds = ( new WikiaSQL() )
+					->SELECT( 'ug_user' )
+					->FROM ( 'user_groups' )
+					->WHERE( 'ug_group' )->IN( [ 'staff', 'util', 'helper', 'vstf' ] )
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: These groups shouldn't be in the local table. Is this query necessary?

**Comment by d34th4ck3r** [comment]
Made changes according to your comments. Please recheck.

File: extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php [hunk 3]
```diff
@@ -190,6 +191,36 @@ function () {
...
+					->FROM ( 'user_groups' )
+					->WHERE( 'ug_group' )->IN( [ 'bot', 'bot-global', 'staff', 'util', 'helper', 'vstf' ] )
+					->GROUP_BY( 'ug_user' )
+					->runLoop( $globalDb, function ( &$globalIds, $row ) {
+						$globalIds[] = $row->ug_user;
+					} );
+
+				$allBlockedIds = array_merge( $globalIds, $this->getBotIds() );
+
+				return $allBlockedIds;
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: This is still using the "blocked" in the variable name, but I'd suggest just changing it to:

``` php
return array_merge( $globalIds, $this->getBotIds() );
```

File: extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php [hunk 4]
```diff
@@ -7,6 +7,7 @@ class CommunityPageSpecialUsersModel {
 	const ALL_ADMINS_MCACHE_KEY = 'community_page_all_admins';
 	const GLOBAL_BOTS_MCACHE_KEY = 'community_page_global_bots';
 	const ALL_BOTS_MCACHE_KEY = 'community_page_all_bots';
+	const ALL_BLOCKED_IDS_MCACHE_KEY = 'community_page_all_blocked_ids';
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: Likewise, still using "Blocked".

**Comment by Grunny** [comment]
Last two comments, otherwise LGTM. :)

File: extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php [hunk 5]
```diff
@@ -190,6 +191,48 @@ function () {
...
+					->runLoop( $globalDb, function ( &$globalIds, $row ) {
+						$globalIds[] = $row->ug_user;
+					} );
+
+				$localDb = wfGetDB( DB_SLAVE );
+
+				$localUsers = ( new WikiaSQL() )
+					->SELECT( 'ug_user' )
+					->FROM ( 'user_groups' )
+					->WHERE( 'ug_group' )->NOT_IN( [ 'bot', 'bot-global' ] )
```
Comment (Grunny) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: Isn't this just [the exact query from `getBotIds`](https://github.com/Wikia/app/blob/6da13d7280cb58024838318061bc83933cd2d6f5/extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php#L174-L185), so would be redundant to calling `getBotIds` below, right?

Comment (d34th4ck3r) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: It's actually the exact opposite. (Notice NOT_IN)
Basically I'm trying to remove local users from blocked ids if they are in any of the global groups.

Comment (Grunny) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: Oh, bah, I missed that. ;)

So, you want to remove local users (who are also one of staff, helper, etc. globally) from being blacklisted if they have local rights (basically if they either have rollback or sysop rights)?

Also, you can remove `bot-global` from this query, that's only a global right, so shouldn't be in the local table.

Comment (d34th4ck3r) [inline extensions/wikia/CommunityPage/models/CommunityPageSpecialUsersModel.class.php]: Right! I want to remove any local users from blacklist unless they are bot.

Done.

**Comment by Grunny** [comment]
LGTM. :)

**Pull request merged**